### PR TITLE
[SPARK-27581][SQL] DataFrame countDistinct("*") shouldn't fail with AnalysisException

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Currently `countDistinct("*")` doesn't work. An analysis exception is thrown:

```scala
val df = sql("select id % 100 from range(100000)")
df.select(countDistinct("*")).first()

org.apache.spark.sql.AnalysisException: Invalid usage of '*' in expression 'count';
```

Users need to use `expr`.

```scala
df.select(expr("count(distinct(*))")).first()
```

This limits some API usage like `df.select(count("*"), countDistinct("*))`.

The PR takes the simplest fix that lets analyzer expand star and resolve `count` function.

## How was this patch tested?

Added unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -358,9 +358,10 @@ object functions {
    * @since 1.3.0
    */
   @scala.annotation.varargs
-  def countDistinct(expr: Column, exprs: Column*): Column = {
-    withAggregateFunction(Count.apply((expr +: exprs).map(_.expr)), isDistinct = true)
-  }
+  def countDistinct(expr: Column, exprs: Column*): Column =
+    // For usage like countDistinct("*"), we should let analyzer expand star and
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: how do we handle it in `count`?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: `count("*")` doesn't go into Analyzer. It is transformed to `count(1)` in `AstBuilder` and `functions` before analysis.

https://github.com/apache/spark/blob/0407070945fc090c84dde7a84e1b24a1acb131d4/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L1363-L1370

https://github.com/apache/spark/blob/90085a184797f8bddbff8ca6ec7a60f3899c1a86/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L337-L343

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**